gpswearos/src/main/java/com/android/gpswearos/presentation/SampleData.kt

```diff
@@ -0,0 +1,15 @@
+package com.android.gpswearos.presentation
+
+object SampleData {
+    // Sample conversation data
+    val statusSample = listOf(
+        SatelliteStatus("Lat:", "28.0713003 °"),

```

Comment:
These fields will all be derived for a `Location` object. SatelliteStatus will hold the info from the GnssStatus object. So I would change this container class to LocationRow of something similar.

Also, you should be able to move some of this common code and resources into a shared module. SatelliteStatus could go there, as well as the strings for lat, lon, etc.

Reply:
Ok, thanks

Reply:
> These fields will all be derived from a `Location` object. SatelliteStatus will hold the info from the GnssStatus object. So I would change this container class to LocationLabelAndData or something similar.
> 
> Also, you should be able to move some of this common code and resources into a shared module. SatelliteStatus could go there, as well as the strings for lat, lon, etc.

Ok, thanks!

Reply:
> > These fields will all be derived from a `Location` object. SatelliteStatus will hold the info from the GnssStatus object. So I would change this container class to LocationLabelAndData or something similar.
> > Also, you should be able to move some of this common code and resources into a shared module. SatelliteStatus could go there, as well as the strings for lat, lon, etc.
> 
> Ok, thanks!

I've created a library module to put sharing data. I've added the module dependency successfully to `wear` module but failed to add it to GPSTest. It says "No build type in module ':library' matches build type 'android'.". Since I haven't figured out how to solve it, I will just apply this to `wear` in the next commit.

Reply:
Ok, thanks. Could you work on this a bit more to see if you can get it to work for the app too? Otherwise there is going to be a lot of code duplication between wear and the app once we get into listening for GNSS updates.

See this for guidance - https://developer.android.com/studio/projects/android-library

Reply:
Once you get the library set up, you can move all the classes in the normal app `.di`, `.data`, `.model`, and `.util` packages to the library (keep them in the respective subpackages), as those will be used to get live GNSS data on Wear (and in the app).

So you should end up having in the library:
* `com.android.gpstest.library.di`
* `com.android.gpstest.library.data`
* `com.android.gpstest.library.model`
* `com.android.gpstest.library.util`

Reply:
> So I worked on this a bit, and it turns out the app just needs to be updated in a few places due to other changes. Here's a working branch showing the data structure in the library being rendered in the normal app Status screen under the LocationCard: https://github.com/barbeau/gpstest/tree/wear-working-library
> 
> Here's the commit with the changes I made to get the library to work with the app: [3ec9aff](https://github.com/barbeau/gpstest/commit/3ec9aff1dda1d5e781108ebf91ab25bd359e1349)
> 
> You can ignore the changes made to the wear MainActivity - those were part of another test I was doing and I accidentally included them in the commit.

Thanks! It works now!

Reply:
> Once you get the library set up, you can move all the classes in the normal app `.di`, `.data`, `.model`, and `.util` packages to the library (keep them in the respective subpackages), as those will be used to get live GNSS data on Wear (and in the app).
> 
> So you should end up having in the library:
> 
> * `com.android.gpstest.library.di`
> * `com.android.gpstest.library.data`
> * `com.android.gpstest.library.model`
> * `com.android.gpstest.library.util`

I've followed this instruction and found some problems. I've moved the `.di`, `.data`, `.model`, and `.util` packages to the `library` at first, and found these use a lot of things in `res/values`. After copying the `res/values` to the current `library`, I found I have to move `Application` to `library` for the same reason. After this,  there are still some classes needed to move to `library`, After moving, I found more classes need to be moved since the new moved classes cite the classes in the origin package. Since I haven't figured out how to solve this, I will first commit the revisions under resolved conversations.

Reply:
>I found I have to move Application to library for the same reason

Ah, yes, good point. So IIRC the Application class is mainly used throughout the app as a way to access the application-wide context to get strings from resources, etc. (as opposed to using the Activity context) and to access a centralized preferences object.

Here is the approach I would recommend:
1. For all the classes in the `.data` package (*Manager.kt), I would replace the reference to `Application.app` with the `context` that is passed into the constructor
2. For all the classes in the `.util` package that use `Application.Companion.getPrefs()` (in Java classes) or `Application.prefs` (in Kotlin classes), I would refactor the methods that use them to take in a `SharedPreferences` as a new parameter.

Then, the wear application and mobile application can both implement their own version of the `Application` class (which will be very similar), and pass in the `Application.app` for the context and `Application.prefs` for the SharedPreferences to all the relevant methods.

If you follow this approach, are there any other classes in GPSTest module that are still needed in the library?

Reply:
> > I found I have to move Application to library for the same reason
> 
> Ah, yes, good point. So IIRC the Application class is mainly used throughout the app as a way to access the application-wide context to get strings from resources, etc. (as opposed to using the Activity context) and to access a centralized preferences object.
> 
> Here is the approach I would recommend:
> 
> 1. For all the classes in the `.data` package (*Manager.kt), I would replace the reference to `Application.app` with the `context` that is passed into the constructor
> 2. For all the classes in the `.util` package that use `Application.Companion.getPrefs()` (in Java classes) or `Application.prefs` (in Kotlin classes), I would refactor the methods that use them to take in a `SharedPreferences` as a new parameter.
> 
> Then, the wear application and mobile application can both implement their own version of the `Application` class (which will be very similar), and pass in the `Application.app` for the context and `Application.prefs` for the SharedPreferences to all the relevant methods. So, with this approach you shouldn't need the Application class in the library.
> 
> If you follow this approach, are there any other classes in GPSTest module that are still needed in the library?

Should the function `sendEmail()` inside `util.UIUtils.kt` be also moved to the library? `SignalInfoViewModel` was cited as a parameter of `sendEmail()`. If 'sendEmail()' was reserved, `SignalInfoViewModel` should also be moved to the library.

Reply:
> > I found I have to move Application to library for the same reason
> 
> Ah, yes, good point. So IIRC the Application class is mainly used throughout the app as a way to access the application-wide context to get strings from resources, etc. (as opposed to using the Activity context) and to access a centralized preferences object.
> 
> Here is the approach I would recommend:
> 
> 1. For all the classes in the `.data` package (*Manager.kt), I would replace the reference to `Application.app` with the `context` that is passed into the constructor
> 2. For all the classes in the `.util` package that use `Application.Companion.getPrefs()` (in Java classes) or `Application.prefs` (in Kotlin classes), I would refactor the methods that use them to take in a `SharedPreferences` as a new parameter.
> 
> Then, the wear application and mobile application can both implement their own version of the `Application` class (which will be very similar), and pass in the `Application.app` for the context and `Application.prefs` for the SharedPreferences to all the relevant methods. So, with this approach you shouldn't need the Application class in the library.
> 
> If you follow this approach, are there any other classes in GPSTest module that are still needed in the library?

Should the function `sendEmail()` inside `util.UIUtils.kt` be also moved to the library? `SignalInfoViewModel` was referenced as a parameter of `sendEmail()`. If 'sendEmail()' was reserved, `SignalInfoViewModel` should also be moved to the library.

Reply:
I believe `SignalInfoViewModel` should work on Wear, so I'd say yes, let's go ahead and move it over. It will help processing data from the GNSS listeners and preparing it for the UI (it powers the Status view in the normal app).

Reply:
> 

`ShareDialogFragment` in `gpstest.ui.share` was used in `UIUtils.kt`. Should code about Fragment be moved out of `UIUtils.kt` and put back to `gpstest`?

Reply:
> `ShareDialogFragment` in `gpstest.ui.share` was used in `UIUtils.kt`. Should code about Fragment be moved out of `UIUtils.kt` and put back to `gpstest`?

`ShareDialogFragment` in `gpstest.ui.share` was used in `UIUtils.kt`. Should code about Fragment be moved out of `UIUtils.kt` and put back to `gpstest`?

Reply:
> I believe `SignalInfoViewModel` should work on Wear, so I'd say yes, let's go ahead and move it over. It will help processing data from the GNSS listeners and preparing it for the UI (it powers the Status view in the normal app).

`ShareDialogFragment` in `gpstest.ui.share` was used in `UIUtils.kt`. Should code about Fragment be moved out of `UIUtils.kt` and put back to `gpstest`?

Reply:
> I believe `SignalInfoViewModel` should work on Wear, so I'd say yes, let's go ahead and move it over. It will help processing data from the GNSS listeners and preparing it for the UI (it powers the Status view in the normal app).

`GnssFilterDialog` and `HelpActivity` also seem shouldn't be shared in `library`. Should I move these to `gpstest`? And should `res` be copied to `library`? The content in `res` has been referenced many times in the newly moved library content.

Reply:
Yes, Dialogs, Fragments, and Activities should stay in the app. For the util methods that use these, they should stay in the utility method class in the app.

Reply:
And yes, at least do_not_translate.xml, and likely some other strings, and possibly some other res (flag images) should move to library.

Reply:
> And yes, at least do_not_translate.xml, and likely some other strings, and possibly some other res (flag images) should move to library.

Thanks. Now, some resources are in GPSTest, Others are in library. However, some codes in library can't find the resources even if the resources are in library too. At first, the compiler couldn't recognize the resources in library and marked this with red line. I've added `xmlns:tools="http://schemas.android.com/tools" tools:ignore="ExtraTranslation"` inside resource tag, and the red line disappeared. The current problem is that the compiler could find the resources but the Gradle couldn't. When I try to build the project, the build results show "Unresolved reference ..."

Reply:
You'll need to make sure that for each string from values/strings.xml and array from values/arrays.xml moved to the library, the corresponding strings are moved for all the translations in the `values-XX` (e.g., `values-es`) folders. In other words, each translation of a string must be in the same project as the English string. If you do this correctly you shouldn't need the `ExtraTranslation` tag.

>the compiler could find the resources but the Gradle couldn't

Could you explain more about this? By "compiler" do you mean Android Studio?

Please try to move all the strings in the right place, and if you're still having problems please push what you have to GitHub and I can take a look.

Reply:
> You'll need to make sure that for each string from values/strings.xml and array from values/arrays.xml moved to the library, the corresponding strings are moved for all the translations in the `values-XX` (e.g., `values-es`) folders. In other words, each translation of a string must be in the same project as the English string. If you do this correctly you shouldn't need the `ExtraTranslation` tag.
> 
> > the compiler could find the resources but the Gradle couldn't
> 
> Could you explain more about this? By "compiler" do you mean Android Studio?
> 
> Please try to move all the strings in the right place, and if you're still having problems please push what you have to GitHub and I can take a look.

Thanks a lot. I've selected the Android tab to view the whole project structure, which caused me to move these XML files to the wrong packages. Now I've solved this problem and selected the Project tab.

Reply:
As we discussed before, I added `SharedPreferences` to the constructors of a few classes. In order to make dependency-injection available, I've created another `@Provides`-annotated method in DataModule.kt to instantiate the required `SharedPreferences`. That method requires a `Context`, which used to be available to `DataModule` already. But the toolchain says that I'll need another `@Provides`-annotated method to create the required `Context`. In the repository before my change, there's no `@Provide`-annotated method that returns `Context` at all. The application context is just accessible directly.  I'm not sure if it's because I moved `DataModule` into library module so it lost such access. I've commit the new version and need your advice to do such cross-module dependency injections. Thanks!

Reply:
@adaext Nice work! Look like you just need to add a `@Provides` method for the context itself in the `DataModule` method - like this:

```kotlin
    @Provides
    fun provideContext(
        @ApplicationContext context: Context,
    ): Context {
        return context
    }
```

I made that change and reduced `targetSdkVersion` to `29` in this commit and it seems to be working:
https://github.com/barbeau/gpstest/tree/wear-provide-context-example

Here's the commit with the changes - https://github.com/barbeau/gpstest/commit/ea8adbcf034b3419984ec514355aa1921f4ca4a4

Reply:
> 

Thanks, it finally works. I reduced the `targetSdkVersion` to 31 because of the error from Gradle, "Google Play requires that apps target API level 31 or higher.". And it also works.

Reply:
You'll need to keep `targetSdkVersion 29` for the app to work on Android 13 devices (for now), otherwise it will crash on an error for the PendingIntent `Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE`.